Source/Scene/PointPrimitiveCollection.js

```diff
@@ -0,0 +1,976 @@
+        var proj = Cartesian3.multiplyByScalar(camera.directionWC, Cartesian3.dot(toCenter, camera.directionWC), scratchProj);
+        var distance = Math.max(0.0, Cartesian3.magnitude(proj) - boundingVolume.radius);
+
+        scratchDrawingBufferDimensions.x = context.drawingBufferWidth;
+        scratchDrawingBufferDimensions.y = context.drawingBufferHeight;
+        var pixelSize = frustum.getPixelSize(scratchDrawingBufferDimensions, distance);
+        var pixelScale = Math.max(pixelSize.x, pixelSize.y);
+
+        var size = pixelScale * collection._maxPixelSize;
+        boundingVolume.radius += size;
+    }
+
+    /**
+     * @private
+     */
+    PointPrimitiveCollection.prototype.update = function(context, frameState, commandList) {
+        removePointPrimitives(this);
+
+        this._maxTotalPointSize = context.maximumAliasedPointSize;
+
+        var pointPrimitives = this._pointPrimitives;
+        var pointPrimitivesLength = pointPrimitives.length;
+
+        updateMode(this, frameState);
+
+        pointPrimitives = this._pointPrimitives;
+        pointPrimitivesLength = pointPrimitives.length;
+        var pointPrimitivesToUpdate = this._pointPrimitivesToUpdate;
+        var pointPrimitivesToUpdateLength = this._pointPrimitivesToUpdateIndex;
+
+        var properties = this._propertiesChanged;
+
+        var createVertexArray = this._createVertexArray;
+
+        var vafWriters;
+        var pass = frameState.passes;
+        var picking = pass.pick;
+
+        // PERFORMANCE_IDEA: Round robin multiple buffers.
+        if (createVertexArray || (!picking && this.computeNewBuffersUsage())) {
+            this._createVertexArray = false;
+
+            for (var k = 0; k < NUMBER_OF_PROPERTIES; ++k) {
+                properties[k] = 0;
+            }
+
+            this._vaf = this._vaf && this._vaf.destroy();
+
+            if (pointPrimitivesLength > 0) {
+                // PERFORMANCE_IDEA:  Instead of creating a new one, resize like std::vector.
+                this._vaf = createVAF(context, pointPrimitivesLength, this._buffersUsage);
+                vafWriters = this._vaf.writers;
+
+                // Rewrite entire buffer if pointPrimitives were added or removed.
+                for (var i = 0; i < pointPrimitivesLength; ++i) {
+                    var pointPrimitive = this._pointPrimitives[i];
+                    pointPrimitive._dirty = false; // In case it needed an update.
+                    writePointPrimitive(this, context, vafWriters, pointPrimitive);
+                }
+
+                // Different pointPrimitive collections share the same index buffer.
+                this._vaf.commit();
+            }
+
+            this._pointPrimitivesToUpdateIndex = 0;
+        } else {
+            // PointPrimitives were modified, but none were added or removed.
+            if (pointPrimitivesToUpdateLength > 0) {
+                var writers = [];
```

Comment:
Ouch, look at this memory allocation.  Can you use a scratch here and in `BillboardCollection`?  Just set its length to zero each time.